packages/gatsby/src/query/index.js

```diff
@@ -38,54 +36,62 @@ const popExtractedQueries = () => {
 }
 
 const findIdsWithoutDataDependencies = state => {
-  const allTrackedIds = _.uniq(
-    _.flatten(
-      _.concat(
-        _.values(state.componentDataDependencies.nodes),
-        _.values(state.componentDataDependencies.connections)
-      )
-    )
+  const allTrackedIds = new Set(
+    _.flatten([
+      ...state.componentDataDependencies.nodes.values(),
+      ...state.componentDataDependencies.connections.values(),
+    ])
   )
 
   // Get list of paths not already tracked and run the queries for these
   // paths.
-  const notTrackedIds = _.difference(
+  const notTrackedIds = new Set(
     [
       ...Array.from(state.pages.values(), p => p.path),
       ...[...state.staticQueryComponents.values()].map(c => c.id),
-    ],
-    [...allTrackedIds, ...seenIdsWithoutDataDependencies]
+    ].filter(
+      x => !allTrackedIds.has(x) && !seenIdsWithoutDataDependencies.has(x)
+    )
   )
 
   // Add new IDs to our seen array so we don't keep trying to run queries for them.
   // Pages without queries can't be tracked.
-  seenIdsWithoutDataDependencies = _.uniq([
-    ...notTrackedIds,
-    ...seenIdsWithoutDataDependencies,
-  ])
+  for (const notTrackedId of notTrackedIds) {
+    seenIdsWithoutDataDependencies.add(notTrackedId)
+  }
 
   return notTrackedIds
 }
 
 const popNodeQueries = state => {
   const actions = _.uniq(queuedDirtyActions, a => a.payload.id)
-  const uniqDirties = _.uniq(
-    actions.reduce((dirtyIds, action) => {
-      const node = action.payload
-
-      if (!node || !node.id || !node.internal.type) return dirtyIds
-
-      // Find components that depend on this node so are now dirty.
-      dirtyIds = dirtyIds.concat(state.componentDataDependencies.nodes[node.id])
+  const uniqDirties = actions.reduce((dirtyIds, action) => {
+    const node = action.payload
+
+    if (!node || !node.id || !node.internal.type) return dirtyIds
+
+    // Find components that depend on this node so are now dirty.
+    if (state.componentDataDependencies.nodes.has(node.id)) {
+      state.componentDataDependencies.nodes.get(node.id).forEach(n => {
+        if (n) {
```

Comment:
This is interesting - did you see some falsy entries in there that you needed to add this check?

Reply:
If this is the case then there is some bug somewhere:
- it can be be not problematic (i.e. we try to add dependency from running queries in gatsby-node, which will never have query identifier and we should just not add dependencies in those cases) 
- or might be very problematic if page/static queries don't add dependency properly in some cases, and so we don't properly invalidate query results for them

Reply:
It's not in the scope of this PR to fix this, but it's nice that it potentially encovered problem

Reply:
Yeah I was getting an error that undefined was being added into the Set. It’s possible it was caused by another issue I introduced and later fixed while creating the PR. I’ll run it without this line in the morning.

Reply:
Yeah seems it now works without it - I'll remove it.

Reply:
Nvm seems the E2E tests fail with it removed :^)

Reply:
Before this was fine as arrays accept undefined, and the _.compact was clearing them out. It seems this issue must've been encountered in the past, hence why _.compact was added in the first place.